File: 322. Coin Change/main.cpp
#include <limits>
#include <vector>

using namespace std;

class Solution {
public:
    int coinChange(vector<int> &coins, int amount) {
        vector<int> dp(amount + 1, numeric_limits<int>::max());
        dp[0] = 0;
        int coins_size = coins.size();

        for (int i = 0; i < amount; i++) {
            if (dp[i] == numeric_limits<int>::max()) {
                continue;
            }

            for (int j = 0; j < coins_size; j++) {
                if (coins[j] <= amount - i) {
                    if (dp[i + coins[j]] > dp[i] + coins[j]) {
                        dp[i + coins[j]] = dp[i] + coins[j];
                    }
                }
            }
        }

        if (dp[amount] == numeric_limits<int>::max()) {
            return -1;
        }

        return dp[amount];
    }
};

class Solution {
public:
    int coinChange(vector<int> &coins, int amount) {
        vector<int> dp(amount + 1, numeric_limits<int>::max());
        dp[0] = 0;
        int coins_size = coins.size();

        for (const auto &coin : coins) {
            for (int i = coin; i <= amount; i++) {
                if (dp[i - coin] == numeric_limits<int>::max()) {
                    continue;
                }
                dp[i] = min(dp[i], dp[i - coin] + 1);
            }
        }

        if (dp[amount] == numeric_limits<int>::max()) {
            return -1;
        }

        return dp[amount];
    }
};

File: 322. Coin Change/main.rs
struct Solution;

impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let amount = amount as usize;
        let mut dp = vec![i32::MAX; amount + 1];
        dp[0] = 0;

        for i in 0..amount {
            if dp[i] == i32::MAX {
                continue;
            }

            for j in 0..coins.len() {
                if coins[j] as usize <= amount - i {
                    dp[i + coins[j] as usize] = dp[i + coins[j] as usize].min(dp[i] + 1);
                }
            }
        }

        if dp[amount] == i32::MAX {
            return -1;
        }

        dp[amount]
    }
}

impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let amount = amount as usize;
        let mut dp = vec![i32::MAX; amount + 1];
        dp[0] = 0;

        for coin in coins {
            for i in coin as usize..=amount {
                if dp[i - coin as usize] == i32::MAX {
                    continue;
                }

                dp[i] = dp[i].min(dp[i - coin as usize] + 1);
            }
        }

        if dp[amount] == i32::MAX {
            return -1;
        }

        dp[amount]
    }
}
